README
This is FLVMeta, a metadata injection and manipulation tool for Adobe(r) FLV video files.

The package contains an utility, flvmeta, to be used to inject the onMetaData event
in FLV files as well as the onLastSecond event, to print all informations about
an input FLV file, as well as to check a file for validity.

The flvmeta utility is able to compute almost all known metadata tags,
including the keyframes list.
Since version 1.0.7, it has also been able to fix very large FLV files with
invalid 24-bit only timestamps to make proper use of 32-bit extended
timestamps.
Since version 1.0.8, it has been able to handle arbitrarily large files.

This project was started as an alternative to tools like flvtool2.
It is written in portable C, and should compile on most platforms (it has
been successfully compiled and tested on Ubuntu Feisty and later, Debian Lenny,
MacOSX 10.4, Windows XP, and Windows 7).
It also cleanly compiles as a 64-bit binary without alteration.
FLVMeta is fast (only bound by I/O speed) and has a very small memory footprint.

The FLVMeta source package includes and uses the following software:
- the libyaml YAML parser and emitter (http://pyyaml.org/wiki/LibYAML)


FLVMeta is provided "as is" with no warranty.  The exact terms
under which you may use and (re)distribute this program are detailed
in the GNU General Public License, in the file COPYING.

See the files AUTHORS and THANKS for a list of authors and other contributors.

See the file INSTALL for compilation and installation instructions.

See the file NEWS for a description of major changes in this release.

See the file TODO for ideas on how you could help us improve FLVMeta.

Send bug reports to flvmeta-discussion@googlegroups.com.


NEWS
Version 1.2.0 (unreleased)
  - Removed support for the autotools build system.
  - Fixed installation instructions for CMake.
  - Switched to Semver 2.0 as versioning scheme.
  - Improved 64-bit and file handling compatibility.
  - Added dumping of AVC and AAC packet information.
  - Improved FLV file checking.  

Version 1.1.2 (2013-08-04)
  - Added JSON as output format for check reports.
  - Fixed automake distribution of the CmakeLists.txt file used
    to build the man page.
  - Added support for the Travis continuous integration platform.
  - Fixed inconsistencies in FLVMeta product spelling.
  - Removed man file from Windows binary packages.

Version 1.1.1 (2013-05-09)
  - Improved file duration detection.
  - Fixed FLVMeta product spelling.
  - Updated copyright notices to 2013.

Version 1.1.0.1 (2012-10-18)
  - No change in the code.
  - Fixed spelling errors in the man page.

Version 1.1.0 (2012-05-03)
  - Beta release. All features should be considered relatively
    stable unless explicitely stated otherwise.
  - Added proper command line handling and help.
  - Added the possibility to overwrite the input file
    when the output file is not specified or when both files
    are physically the same.
  - Added support for CMake builds in addition to autotools.
    It is now the official way to build flvmeta on Windows.
  - Added metadata and full file dumping, integrating former
    flvdump functionality into flvmeta.
  - Added support for XML, YAML, and JSON formats for dumping.
  - Added XML schemas describing the various formats used by flvmeta.
  - Added a file checking feature.
  - Added the possibility to print output file metadata after
    a successful update using one of the supported formats.
  - Added a feature to insert custom metadata strings while updating.
  - Added an option to disable insertion of the onLastSecond event.
  - Added an option to preserve existing metadata tags if possible.
  - Added an option to fix invalid tags while updating (this is
    a highly experimental feature, should be used with caution)
  - Added an option to ignore invalid tags as much as possible
    instead of exiting at the first error detected.
  - Added an option to reset the file timestamps in order to
    correctly start at zero, for example if the file has been
    incorrectly split by buggy tools.
  - Added an option to display informative messages while processing
    (not quite exhaustive for the moment).
  - Changed keyframe index generation so only the first keyframe
    for a given timestamp will be indexed. This behaviour can be
    overriden with the --all-keyframes/-k option.

Version 1.0.11 (2010-01-25)
  - Bug fix release.
  - Fixed video resolution detection when the first video frame
    is not a keyframe.
  - Fixed invalid timestamp handling in the case of decreasing timestamps.
  - Fixed AVC resolution computation when frame cropping rectangle
    is used.
  - Fixed handling of files with a non-zero starting timestamp.
  - Fixed datasize tag computation so only metadata are taken
    into account.

Version 1.0.10 (2009-09-02)
  - Bug fix release.
  - Fixed amf_data_dump numeric format.
  - Fixed extended timestamp handling.
  - Fixed video resolution detection causing a crash in the case
    the video tag data body has less data than required.

Version 1.0.9 (2009-06-23)
  - Bug fix release.
  - Fixed large file support so it will work for files bigger than 4GB.
  - Fixed date handling in AMF according to the official spec.
  - Fixed extended timestamp handling.
  - Fixed a bug causing reading invalid tags to potentially lead to
    memory overflow and creation of invalid files.

Version 1.0.8 (2009-05-08)
  - Added support for arbitrary large files (2GB+).
  - Added support for AVC (H.264) video resolution detection.
  - Added support for the Speex codec and rarely met video frame types.
  - Fixed a bug where two consecutive tags of different types and
    with decreasing timestamps would cause extended timestamps
    to be incorrectly used for the next tags.
  - Fixed a bug where zero would be used as video height and width
    if video resolution could not be detected.
  - Fixed a bug causing flvdump to crash after reading invalid tags.
    Flvdump now exits after the first invalid tag read.

Version 1.0.7 (2008-09-25)
  - Added support for extended timestamps.
    Now flvmeta can read and write FLV files longer than 4:39:37,
    as well as fix files with incorrect timestamps.
  - Added support for all codecs from the official specification.
  - Fixed a bug causing flvdump to lose track of tags in case
    of invalid metadata.

Version 1.0.6 (2008-05-28)
  - Bug fix release.
  - Fixed a flvdump crash under Linux caused by incorrect string
    handling.

Version 1.0.5 (2008-04-03)
  - Bug fix release.
  - Fixed a bug in the video size detection for VP60.
  - Added support for the AMF NULL type.
  - Simplified the AMF parser/writer.

Version 1.0.4 (2008-01-04)
  - Bug fix release.
  - Fixed a bug where some data tags wouldn't be written.
  - Fixed a date computation bug related to daylight saving.
  - Modified flvdump to make it more tolerant to malformed files.

Version 1.0.3 (2007-10-21)
  - Bug fix release.
  - Fixed major bugs in the AMF parser/writer.
  - Fixed a bug in the video size detection for VP6 alpha.
  - Fixed minor bugs.

Version 1.0.2 (2007-09-30)
  - Bug fix release.
  - Fixed issues on 64-bits architectures.
  - Fixed "times" metadata tag, which was incorrectly written
    as "timestamps".
  - Fixed audio delay computation.

Version 1.0.1 (2007-09-25)
  - Bug fix release.
  - Fixed a critical bug where file size and offsets would not
    be correctly computed if the input file did not have
    an onMetaData event.
  - Audio related metadata are not added anymore if the FLV file
    has no audio data.

Version 1.0 (2007-09-21)
  - This is the first public release.


flvdump.c
/*
    $Id$

    FLV Metadata updater

    Copyright (C) 2007-2010 Marc Noirot <marc.noirot AT gmail.com>

    This file is part of FLVMeta.

    FLVMeta is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    FLVMeta is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with FLVMeta; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
#include "flvmeta.h"
#include "flv.h"
#include "amf.h"

#include <stdio.h>
#include <string.h>

int main(int argc, char ** argv) {

    if (argc < 2) {
        fprintf(stderr, "FLVDump version %s - Copyright (C) 2007, 2008, 2009 Marc Noirot\n\n",
            VERSION);
        fprintf(stderr, "Usage: flvdump in_file\n\n");
        fprintf(stderr, "This is free software; see the source for copying conditions. There is NO\n"
                        "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\n");
        fprintf(stderr, "Report bugs to <%s>\n\n", PACKAGE_BUGREPORT);
        return 1;
    }

    FILE * flv_in = fopen(argv[1], "rb");

    if (flv_in == NULL) {
        fprintf(stderr, "error opening %s.\n", argv[1]);
        return 2;
    }

    flv_header fh;

    if (fread(&fh, sizeof(fh), 1, flv_in) == 0) {
        fclose(flv_in);
        fprintf(stderr, "error reading header\n");
        return 3;
    }

    printf("Magic: %.3s\n", fh.signature);
    printf("Version: %d\n", fh.version);
    printf("Has audio: %s\n", flv_header_has_audio(fh) ? "yes" : "no");
    printf("Has video: %s\n", flv_header_has_video(fh) ? "yes" : "no");
    printf("Offset: %u\n", swap_uint32(fh.offset));

    /* skip first empty previous tag size */
    flvmeta_fseek(flv_in, sizeof(uint32_be), SEEK_CUR);

    flv_tag ft;
    file_offset_t offset;
    char * str;
    uint32 n = 1;
    uint32 prev_tag_size;
    
    /* extended timestamp handling */
    uint32 prev_timestamp_video = 0;
    uint32 prev_timestamp_audio = 0;
    uint32 prev_timestamp_meta = 0;
    uint8 timestamp_extended_video = 0;
    uint8 timestamp_extended_audio = 0;
    uint8 timestamp_extended_meta = 0;
    
    while (!feof(flv_in)) {
        offset = flvmeta_ftell(flv_in);
        if (fread(&ft, sizeof(ft), 1, flv_in) == 0)
            break;

#if defined(_FILE_OFFSET_BITS) && _FILE_OFFSET_BITS == 64
        printf("--- Tag #%u at 0x%llX",  n++, (sint64)offset);
        printf(" (%lli) ---\n", (sint64)offset);
#else
        printf("--- Tag #%u at 0x%lX (%li) ---\n", n++, (long unsigned int)offset, (long int)offset);
#endif
        switch (ft.type) {
            case FLV_TAG_TYPE_AUDIO: str = "Audio"; break;
            case FLV_TAG_TYPE_VIDEO: str = "Video"; break;
            case FLV_TAG_TYPE_META: str = "Meta"; break;
            default: str = "Unknown";
        }
        printf("Tag type: %s\n", str);

        uint32 body_length = uint24_be_to_uint32(ft.body_length);
        printf("Body length: %u\n", body_length);
        
        uint32 timestamp = flv_tag_get_timestamp(ft);
        printf("Timestamp: %u", timestamp);

        if (ft.type == FLV_TAG_TYPE_META) {
            if (timestamp < prev_timestamp_meta) {
                ++timestamp_extended_meta;
            }
            prev_timestamp_meta = timestamp;
            if (timestamp_extended_meta > 0) {
                timestamp += (timestamp_extended_meta << 24);
                printf(" (should be %u)", timestamp);
            }
        }
        else if (ft.type == FLV_TAG_TYPE_AUDIO) {
            if (timestamp < prev_timestamp_audio) {
                ++timestamp_extended_audio;
            }
            prev_timestamp_audio = timestamp;
            if (timestamp_extended_audio > 0) {
                timestamp += (timestamp_extended_audio << 24);
                printf(" (should be %u)", timestamp);
            }
        }
        else if (ft.type == FLV_TAG_TYPE_VIDEO) {
            if (timestamp < prev_timestamp_video) {
                ++timestamp_extended_video;
            }
            prev_timestamp_video = timestamp;
            if (timestamp_extended_video > 0) {
                timestamp += (timestamp_extended_video << 24);
                printf(" (should be %u)", timestamp);
            }
        }

        printf("\n");

        if (ft.type == FLV_TAG_TYPE_AUDIO) {
            flv_audio_tag at;
            if (fread(&at, sizeof(at), 1, flv_in)  == 0) {
                fclose(flv_in);
                fprintf(stderr, "error reading audio data\n");
                return 4;
            }

            switch (flv_audio_tag_sound_type(at)) {
                case FLV_AUDIO_TAG_SOUND_TYPE_MONO: str = "Mono"; break;
                case FLV_AUDIO_TAG_SOUND_TYPE_STEREO: str = "Stereo"; break;
                default: str = "Unknown";
            }
            printf("* Sound type: %s\n", str);

            switch (flv_audio_tag_sound_size(at)) {
                case FLV_AUDIO_TAG_SOUND_SIZE_8: str = "8 bits"; break;
                case FLV_AUDIO_TAG_SOUND_SIZE_16: str = "16 bits"; break;
                default: str = "Unknown";
            }
            printf("* Sound size: %s\n", str);

            switch (flv_audio_tag_sound_rate(at)) {
                case FLV_AUDIO_TAG_SOUND_RATE_5_5: str = "5.5 kHz"; break;
                case FLV_AUDIO_TAG_SOUND_RATE_11: str = "11 kHz"; break;
                case FLV_AUDIO_TAG_SOUND_RATE_22: str = "22 kHz"; break;
                case FLV_AUDIO_TAG_SOUND_RATE_44: str = "44 kHz"; break;
                default: str = "Unknown";
            }
            printf("* Sound rate: %s\n", str);

            switch (flv_audio_tag_sound_format(at)) {
                case FLV_AUDIO_TAG_SOUND_FORMAT_LINEAR_PCM: str = "Linear PCM"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_ADPCM: str = "ADPCM"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_MP3: str = "MP3"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_LINEAR_PCM_LE: str = "Linear PCM, little-endian"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_NELLYMOSER_16_MONO: str = "Nellymoser 16-kHz mono"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_NELLYMOSER_8_MONO: str = "Nellymoser 8-kHz mono"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_NELLYMOSER: str = "Nellymoser"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_G711_A: str = "G.711 A-law logarithmic PCM"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_G711_MU: str = "G.711 mu-law logarithmic PCM"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_RESERVED: str = "reserved"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_AAC: str = "AAC"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_SPEEX: str = "Speex"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_MP3_8: str = "MP3 8-Khz"; break;
                case FLV_AUDIO_TAG_SOUND_FORMAT_DEVICE_SPECIFIC: str = "Device-specific sound"; break;
                default: str = "Unknown";
            }
            printf("* Sound format: %s\n", str);
        }
        else if (ft.type == FLV_TAG_TYPE_VIDEO) {
            flv_video_tag vt;
            if (fread(&vt, sizeof(vt), 1, flv_in)  == 0) {
                fclose(flv_in);
                fprintf(stderr, "error reading video data\n");
                return 5;
            }

            switch (flv_video_tag_codec_id(vt)) {
                case FLV_VIDEO_TAG_CODEC_JPEG: str = "JPEG"; break;
                case FLV_VIDEO_TAG_CODEC_SORENSEN_H263: str = "Sorensen H.263"; break;
                case FLV_VIDEO_TAG_CODEC_SCREEN_VIDEO: str = "Screen Video"; break;
                case FLV_VIDEO_TAG_CODEC_ON2_VP6: str = "On2 VP6"; break;
                case FLV_VIDEO_TAG_CODEC_ON2_VP6_ALPHA: str = "On2 VP6 Alpha"; break;
                case FLV_VIDEO_TAG_CODEC_SCREEN_VIDEO_V2: str = "Screen Video V2"; break;
                case FLV_VIDEO_TAG_CODEC_AVC: str = "AVC"; break;
                default: str = "Unknown";
            }
            printf("* Video codec: %s\n", str);

            switch (flv_video_tag_frame_type(vt)) {
                case FLV_VIDEO_TAG_FRAME_TYPE_KEYFRAME: str = "keyframe"; break;
                case FLV_VIDEO_TAG_FRAME_TYPE_INTERFRAME: str = "inter frame"; break;
                case FLV_VIDEO_TAG_FRAME_TYPE_DISPOSABLE_INTERFRAME: str = "disposable inter frame"; break;
                case FLV_VIDEO_TAG_FRAME_TYPE_GENERATED_KEYFRAME: str = "generated keyframe"; break;
                case FLV_VIDEO_TAG_FRAME_TYPE_COMMAND_FRAME: str = "video info/command frame"; break;
                default: str = "Unknown";
            }
            printf("* Video frame type: %s\n", str);
        }
        else if (ft.type == FLV_TAG_TYPE_META) {
            amf_data * data = amf_data_file_read(flv_in);
            size_t data_size = amf_data_size(data);
            printf("* Metadata event name: %s\n", amf_string_get_bytes(data));
            amf_data_free(data);

            data = amf_data_file_read(flv_in);
            data_size += amf_data_size(data);
            printf("* Metadata contents: ");
            amf_data_dump(stdout, data, 0);
            printf("\n");
            amf_data_free(data);

            if (body_length > data_size) {
                printf("* Garbage: %i bytes\n", (int)(body_length - data_size));
            }
            else if (body_length < data_size) {
                printf("* Missing: %i bytes\n", -(int)(body_length - data_size));
            }
        }
        else {
            break;
        }
        flvmeta_fseek(flv_in, offset + sizeof(flv_tag) + body_length, SEEK_SET);
        if (fread(&prev_tag_size, sizeof(uint32_be), 1, flv_in) == 1) {
            printf("Previous tag size: %u\n", swap_uint32(prev_tag_size));
        }
    }

    fclose(flv_in);
    return 0;
}
